sycl/include/sycl/ext/intel/experimental/host_pipes.hpp

```diff
@@ -0,0 +1,93 @@
+//==---------------- pipes.hpp - SYCL pipes ------------*- C++ -*-----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// ===--------------------------------------------------------------------=== //
+
+#pragma once
+
+#include <sycl/context.hpp>
+#include <sycl/device.hpp>
+#include <sycl/ext/intel/experimental/pipe_properties.hpp>
+#include <sycl/ext/oneapi/properties/properties.hpp>
+#include <sycl/queue.hpp>
+#include <type_traits>
+
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+#include <xpti/xpti_data_types.h>
+#include <xpti/xpti_trace_framework.hpp>
+#endif
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace intel {
+namespace experimental {
+
+using default_pipe_properties =
+    decltype(sycl::ext::oneapi::experimental::properties(min_capacity<0>));
+
+template <class _name, class _dataT, class _propertiesT>
+class
+#ifdef __SYCL_DEVICE_ONLY__
+    [[__sycl_detail__::add_ir_attributes_global_variable("sycl-host-access",
+                                                         "readwrite")]]
+#endif
+    // TODO: change name to pipe, and merge into the existing pipe
+    // implementation
+    host_pipe {
+
+  struct
+// Commented out since the host_pipe attribute is not introduced by the front
+// end yet. Confirm with Rob
+#ifdef __SYCL_DEVICE_ONLY__
+      [[__sycl_detail__::add_ir_attributes_global_variable(
+          "sycl-host-pipe",
+          nullptr)]] [[__sycl_detail__::
+                           host_pipe]] [[__sycl_detail__::
+                                             global_variable_allowed]] // may
+                                                                       // not be
+                                                                       // needed
+#endif
+      __pipeType {
+    const char __p;
+  };
+
+  static constexpr __pipeType __pipe = {0};
+
+public:
+  using value_type = _dataT;
+  static constexpr int32_t min_cap =
+      _propertiesT::template has_property<min_capacity_key>()
+          ? _propertiesT::template get_property<min_capacity_key>().value
+          : 0;
```

Comment:
Does this work correctly when there is no `min_capacity` property in the list? To my knowledge the ternary operator isn't `constexpr` so it will try to evaluate both sides. Alternatively you can make a trait in the `detail` namespace and specialize the value of that to whether or not the property is there.

Reply:
@steffenlarsen I'm working on combining the class in this header into ext::intel::experimental::pipes for Zibai. I'm not sure I understand your trait suggestion, could you elaborate? I guess another option is to use a lambda to evaluate this (as well as the other properties which all need to be passed) but that could be a little unwieldy.

Reply:
You could add something like
```c++
namespace detail {
template <typename Properties, typename PropertyKey, typename Cond = void>
struct ValueOrDefault {
  template <typename ValT> static constexpr ValT get(ValT Default) {
    return Default;
  }
};

template <typename Properties, typename PropertyKey>
struct ValueOrDefault<
    Properties, PropertyKey,
    std::enable_if_t<
        ext::oneapi::experimental::is_property_list_v<Properties> &&
        Properties::template has_property<PropertyKey>()>> {
  template <typename ValT> static constexpr ValT get(ValT) {
    return Properties::template get_property<PropertyKey>().value;
  }
};
} // namespace detail
```
then you can use it here as
```c++
 static constexpr int32_t min_cap = detail::ValueOrDefault<_propertiesT, min_capacity_key>::get<int32_t>(0);
```